lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
+            else:
+                # Just make sure we never re-write one.
+                # TODO: possibly merits a proper Exception with message
+                assert self._existing_dim[dim_name] == length
+
+            # Add new coords (mesh or dim) to the already-seen list.
+            for coord in matching_coords:
+                if coord not in self._dim_names_and_coords.coords:
+                    self._dim_names_and_coords.append(dim_name, coord)
+
+            # Add the latest name to the returned list
+            dimension_names.append(dim_name)
+
+        # Get info on mesh, first.
+        dimension_names.clear()
+        mesh = cube.mesh
+        if mesh is None:
+            cube_mesh_dim = None
+        else:
+            # Identify all the mesh dimensions.
+            # NOTE: one of these will be a cube dimension, but that one does not
+            # get any special handling.  We *do* want to list/create them in a
+            # definite order (face,edge,node), and before non-mesh dimensions.
+            mesh_location_dimnames = {}
+            for location in MESH_LOCATIONS:
+                # Find if this location exists in the mesh, and a characteristic
+                # coordinate to identify it with.
+                # To use only _required_ UGRID components, we use a location
+                # coord for nodes, but a connectivity for faces/edges
+                if location == "node":
+                    # For nodes, identify the dim with a coordinate variable.
+                    # Selecting the X-axis one for definiteness.
+                    dim_coords = mesh.coords(include_nodes=True, axis="x")
                 else:
-                    # Return the dim_name associated with the existing
-                    # coordinate.
-                    dim_name = self._name_coord_map.name(coord)
-                    dimension_names.append(dim_name)
+                    # For face/edge, use the non-optional connectivity variable.
+                    cf_role = f"{location}_node_connectivity"
+                    dim_coords = mesh.connectivities(cf_role=cf_role)
+                if len(dim_coords) > 0:
+                    # As the mesh contains this location, we want to include this
+                    # dim in our returned mesh dims.
+                    # We should have 1 identifying variable (of either type).
+                    assert len(dim_coords) == 1
+                    mesh_coord = dim_coords[0]
+                    match = mesh_coord in self._dim_names_and_coords.coords
+                    if match:
+                        # For mesh-identifying coords, we require the *same*
+                        # coord, not an identical one (i.e. "is" not "==")
+                        name = self._dim_names_and_coords.name(mesh_coord)
+                        stored_coord = self._dim_names_and_coords.coord(name)
+                        match = mesh_coord is stored_coord
```

Comment:
@pp-mo Can you give an example where you would find an *identical* coord and not the *same* one?

Reply:
Well, certainly if we are saving cubes loaded from similar input files, but not the same one.
I'm not sure this is the place to explain that though.  What do you think?

Reply:
No worries, I just asked because I was in that space and was curious... I can appreciate that it's more of a discussion point.

All good, nothing to see here... happily moving along :wink: